src/helpers/worker-hours-info.model.ts

```diff
@@ -100,84 +125,211 @@ export class WorkerHourInfo {
-    return Math.round(this.calculateRequiredHoursFromVerboseDates(dates));
+  private static calculateWorkerOvertime(
+    actualShiftsFromCurrentMonth: ShiftCode[],
+    primaryScheduleWorkerShifts: ShiftCode[],
+    currentMonthDates: DateInformationForWorkInfoCalculation[],
+    workerNorm: number
+  ): number {
+    const workerHourNorm = this.calculateWorkerHourNorm(
+      actualShiftsFromCurrentMonth,
+      primaryScheduleWorkerShifts,
+      currentMonthDates,
+      workerNorm
+    );
+    const workerActualWorkTime = this.calculateWorkerActualWorkTime(actualShiftsFromCurrentMonth);
+
+    const diffBetweenRevisionsOvertime = this.calculateOvertimeForRevisionDifference(
+      actualShiftsFromCurrentMonth,
+      primaryScheduleWorkerShifts
+    );
+    const exceedMaximumDayWorkTimeOvertime = this.calculateOvertimeForExceeding(
+      actualShiftsFromCurrentMonth,
+      primaryScheduleWorkerShifts
+    );
+
+    const normAndActualDiff = workerActualWorkTime - workerHourNorm;
+    const algorithmOvertime = diffBetweenRevisionsOvertime + exceedMaximumDayWorkTimeOvertime;
+
+    return Math.max(normAndActualDiff, algorithmOvertime);
   }
 
-  public static calculateRequiredHoursFromVerboseDates(
-    verboseDates: MonthDataArray<Pick<VerboseDate, "isPublicHoliday" | "dayOfWeek">>
+  private static calculateOvertimeForRevisionDifference(
+    actualShiftsFromCurrentMonth: ShiftCode[],
+    primaryScheduleWorkerShifts: ShiftCode[]
   ): number {
-    const workingDaysCount = verboseDates.filter((d) => VerboseDateHelper.isWorkingDay(d)).length;
-    const holidaySaturdaysCount = verboseDates.filter((d) => VerboseDateHelper.isHolidaySaturday(d))
-      .length;
-    const requiredHours = WORK_HOURS_PER_DAY * (workingDaysCount - holidaySaturdaysCount);
+    const monthShiftsWithHistoryShifts = _.zip(
+      actualShiftsFromCurrentMonth,
+      primaryScheduleWorkerShifts
+    ) as [ShiftCode, ShiftCode][];
 
-    return requiredHours;
+    let diffBetweenRevisionsOvertime = 0;
+    monthShiftsWithHistoryShifts.forEach(([actualShift, historyShift]) => {
+      if (actualShift !== historyShift) {
+        const workHoursDiffBetweenRevisions =
+          ShiftHelper.shiftCodeToWorkTime(SHIFTS[actualShift]) -
+          ShiftHelper.shiftCodeToWorkTime(SHIFTS[historyShift]);
+        diffBetweenRevisionsOvertime +=
+          workHoursDiffBetweenRevisions > 0 ? workHoursDiffBetweenRevisions : 0;
+      }
+    });
+    return diffBetweenRevisionsOvertime;
+  }
+
+  private static calculateOvertimeForExceeding(
+    actualShiftsFromCurrentMonth: ShiftCode[],
+    primaryScheduleWorkerShifts: ShiftCode[]
+  ): number {
+    const monthShiftsWithHistoryShifts = _.zip(
+      actualShiftsFromCurrentMonth,
+      primaryScheduleWorkerShifts
+    ) as [ShiftCode, ShiftCode][];
+
+    let exceedMaximumDayWorkTimeOvertime = 0;
+    monthShiftsWithHistoryShifts
+      .filter(([actualShift, historyShift]) => actualShift === historyShift)
```

Comment:
Dlaczego filter.

Moim zdaniem to, że nadgodziny dla takich zmian są liczone w innym miejscu nie oznacza, że my nie powinniśmy uwzględzniać to, że takie zmiany mogą wynosić ponad 12 godzin.

Na przykład.

Bazowa zmiana:
Z1=23
Aktualna zmiana:
Z2=24

Z tego co widzę, obecnie, dla tych dwóch zmian jako nadgodziny my dostaniemy 1, zamiast 12, które musą być (aktualna zmiana Z2 trwa 24 godziny).

Reply:
Więc my raczej powinniśmy to uwgzlędniać, ale dawać tutaj Max z dwóch możliwych typóœ nadgodzin

Reply:
Słuszna uwaga. Szacun, że na to wpadłeś :1st_place_medal: